atc/atccmd/command.go

```diff
@@ -965,25 +1004,21 @@ func (cmd *RunCommand) constructBackendMembers(
 				},
 				cmd.JobSchedulingMaxInFlight,
 			),
-			cmd.ComponentRunnerInterval,
-		)},
-		{Name: atc.ComponentBuildTracker, Runner: builds.NewRunner(
-			logger.Session("tracker-runner"),
-			clock.NewClock(),
-			builds.NewTracker(
-				logger.Session(atc.ComponentBuildTracker),
-				dbBuildFactory,
-				engine,
-			),
-			cmd.ComponentRunnerInterval,
-			bus,
-			lockFactory,
-			componentFactory,
-		)},
-		// run separately so as to not preempt critical GC
-		{Name: atc.ComponentBuildReaper, Runner: lockrunner.NewRunner(
-			logger.Session(atc.ComponentBuildReaper),
-			gc.NewBuildLogCollector(
+		},
+		{
+			Component: atc.Component{
+				Name:     atc.ComponentBuildTracker,
+				Interval: cmd.BuildTrackerInterval,
+			},
+			Runnable: builds.NewTracker(dbBuildFactory, engine),
+			Drainer:  engine,
```

Comment:
I wonder if it makes sense to compose the Drainer into the Runnable interface. It would be nice for anyone implementing a component to just have to implement the Drain method alongside Run, and not have to know to do this. I'm thinking something along the lines of a RunDrainer (similar to a ReadCloser).

Reply:
I actually had it that way at one point, but we don't have any other cases where a component needs to drain, so it didn't feel worth making all of the interfaces (including `Coordinator` and `Scheduler`) more complicated.

Felt like putting it up here was better so we don't have 3/4 components with no-op `Drain` methods, and this particular functionality is already covered at the integration level anyway. More generally I would prefer to avoid relying on explicit draining too much since you never know when the ATC will get `kill -9`ed, and it often leads to annoying slowness when shutting down the `web` nodes.

Reply:
Each component wouldn't need to have a no-op implementation of Drain. Instead we could wrap each in a NoopRunDrainer whenever it needs to fulfill that interface.

Reply:
Ahhh I see, you're not saying to add `Drain` to the `Runnable` interface, you're saying we can add another interface which components can conform to and if they do we'll do the draining automatically. Yeah I'll play with that!